DS-4398 initial implementation of the resource policies endpoint

Micheleboychuk opened the pull request:
This PR addresses the first part of the issue https://jira.lyrasis.org/browse/DS-4398
Contract: https://github.com/DSpace/Rest7Contract/pull/87

DONE
- sync implementation of findAll with the contract
- method to create a resource policy
- method to patch a resource policy
- method to delete a resource policy
- search methods for the resourcepolicies
- enforce security for all the implemented methods
- Integration Tests (full coverage)

Please note that this PR includes the commit from @AlexanderS PR https://github.com/DSpace/DSpace/pull/2447

Postponed to a new PR
- cleanup of ResourcePolicyRest groupUUID and epersonUUID attributes used in the submission area

Commit:
DS-4275: Fix case of resourcepolicies endpoint

The endpoint is specified in lowercase in the REST contract.

Micheleboychuk committed:
DS-4398 initial implementation of the resource policies endpoint

DONE
- sync implementation of findAll with the contract
- add ITs for implemented methods

TODO
- enforce security in access resource policy details
- method to create a resource policy
- method to patch a resource policy

Micheleboychuk committed:
Merge branch 'master' of https://github.com/DSpace/DSpace into resourcepolicies

# Conflicts:
#	dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java

Micheleboychuk committed:
DS-4398 fix checkstyle issues

Micheleboychuk committed:
I added the following unit tests:
- findOneForbiddenTest()
- searchResoucesPolicyByEPersonUuidForbiddenTest()
- searchResoucesPoliciesByResourceUuidForbiddenTest()
- searchResoucesPolicyByGroupUuidForbiddenTest()

Micheleboychuk committed:
created:
    class:
        - resourcePolicyRestPermissionEvaluatorPlugin
    method:
       - boolean hasDSpasePermission(Authentication authentication, Serializable targetId, String targetType, DSpaceRestPermission permission)
       - boolean isMyResoursePolicy(Context context, EPerson eperson, Integer id)  - this one checks if the resourcePolicy, that I'm searching, is of the authenticated person
       - ResourcePolicy findOneById (DAO)

abollini commented on dspace-api/src/main/java/org/dspace/authorize/dao/ResourcePolicyDAO.java:
```diff
@@ -79,4 +80,33 @@
      */
     public List<ResourcePolicy> findByDSoAndActionExceptRpType(Context c, DSpaceObject o, int actionID,
             String rpType) throws SQLException;
+
```
@Micheleboychuk our contribution rules require javadoc for all the new public methods. Please add here and in all the other public methods in this interface

abollini commented on dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:
```diff
@@ -203,32 +202,147 @@ public void deleteByDsoAndTypeNotEqualsTo(Context context, DSpaceObject dso, Str
+    }
+
+    @Override
+    public List<ResourcePolicy> findbyEPerson(Context context, EPerson ePerson, int offset, int limit)
```
can you fix the camelCase please? I mean `findByEPerson`

abollini commented on dspace-api/src/main/java/org/dspace/authorize/service/ResourcePolicyService.java:
```diff
@@ -66,26 +67,60 @@
     public List<ResourcePolicy> findExceptRpType(Context c, DSpaceObject o, int actionID, String rpType)
             throws SQLException;
+
+    public List<ResourcePolicy> findByEPerson(Context c, EPerson ePerson, int offset, int limit) throws SQLException;
```
please add the javadoc here and in the following new methods

abollini commented on dspace-api/src/main/java/org/dspace/authorize/service/ResourcePolicyService.java:
```diff
@@ -66,26 +67,60 @@
 
     public void removeDsoGroupPolicies(Context context, DSpaceObject dso, Group group)
-        throws SQLException, AuthorizeException;
+            throws SQLException, AuthorizeException;
```
there are many small formatting changes in the following lines that are not necessary. It would be great if you can undo them

abollini commented on dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:
```diff
@@ -109,31 +114,25 @@ protected ResourcePolicyDAOImpl() {
-                                       criteriaBuilder.notEqual(resourcePolicyRoot.get(ResourcePolicy_.id), notPolicyID)
-                   )
-        );
+        criteriaQuery.where(
```
please undo formatting changes

abollini commented on dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:
```diff
@@ -109,31 +114,25 @@ protected ResourcePolicyDAOImpl() {
-                                        (resourcePolicyRoot.get(ResourcePolicy_.epersonGroup).in(groups)))
-            )
-        );
+        criteriaQuery.where(criteriaBuilder.and(
```
please undo formatting changes

abollini commented on dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:
```diff
@@ -203,32 +202,147 @@ public void deleteByDsoAndTypeNotEqualsTo(Context context, DSpaceObject dso, Str
-                                    )
-                )
-            );
+            criteriaQuery.where(criteriaBuilder.and(
```
please undo formatting changes

abollini commented on dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:
```diff
@@ -203,32 +202,147 @@ public void deleteByDsoAndTypeNotEqualsTo(Context context, DSpaceObject dso, Str
-                            criteriaBuilder.isNotNull(resourcePolicyRoot.get(ResourcePolicy_.rptype))
-                    )
-            );
+            criteriaQuery.where(criteriaBuilder.and(
```
please undo formatting changes

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +72,90 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
+        return converter.toRestPage(resourcePolisies, pageable, total, utils.obtainProjection(true));
+    }
+
+    @PreAuthorize("hasAuthority('AUTHENTICATED')")
```
due to https://jira.lyrasis.org/browse/DS-4199 this will be ineffective. In any case at this stage (in this PR) your are not yet testing the security constraints stated in the contract so it is fine for me

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +72,90 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
+        return converter.toRestPage(resourcePolisies, pageable, total, utils.obtainProjection(true));
+    }
+
+    @PreAuthorize("hasAuthority('AUTHENTICATED')")
```
due to https://jira.lyrasis.org/browse/DS-4199 this will be ineffective. In any case at this stage (in this PR) your are not yet testing the security constraints stated in the contract so it is fine for me

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +72,90 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
         return ResourcePolicyRest.class;
     }
+
+    @PreAuthorize("hasAuthority('AUTHENTICATED')")
```
due to https://jira.lyrasis.org/browse/DS-4199 this will be ineffective. In any case at this stage (in this PR) your are not yet testing the security constraints stated in the contract so it is fine for me

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+
+        ResourcePolicy resourcePolicyOfCollection = ResourcePolicyBuilder.createResourcePolicy(context)
+                .withDspaceObject(collection).withAction(Constants.REMOVE)
+                .withPolicyType(ResourcePolicy.TYPE_SUBMISSION).withUser(eperson1).build();
```
usually resourcepolicy that apply to a collection haven't a type so it would be better use a more "real-world" example here removing the policyType

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+        ResourcePolicy resourcePolicy = ResourcePolicyBuilder.createResourcePolicy(context)
+                .withDspaceObject(community)
+                .withAction(Constants.READ)
+                .withPolicyType(ResourcePolicy.TYPE_CUSTOM)
```
also resource policies of communities usually haven't a policy type, please remove

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+        ResourcePolicy firstResourcePolicyOfEPerson1 = ResourcePolicyBuilder.createResourcePolicy(context)
+                .withDspaceObject(community)
+                .withAction(Constants.WRITE)
+                .withPolicyType(ResourcePolicy.TYPE_CUSTOM)
```
better unset the policy type

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+        ResourcePolicy firstResourcePolicyOfEPerson2 = ResourcePolicyBuilder.createResourcePolicy(context)
+                .withDspaceObject(community)
+                .withAction(Constants.ADD)
+                .withPolicyType(ResourcePolicy.TYPE_CUSTOM)
```
better unset the policy type

abollini replied:
ok, now we have two resourcepolicies one with and one without the policy type +1

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+        ResourcePolicy firstResourcePolicyOfEPerson1 = ResourcePolicyBuilder.createResourcePolicy(context)
+                .withDspaceObject(collection)
+                .withAction(Constants.WRITE)
+                .withPolicyType(ResourcePolicy.TYPE_CUSTOM)
```
better unset the policy type

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+        ResourcePolicy secondResourcePolicyOfEPerson1 = ResourcePolicyBuilder.createResourcePolicy(context)
+                .withDspaceObject(collection)
+                .withAction(Constants.ADD)
+                .withPolicyType(ResourcePolicy.TYPE_CUSTOM)
```
better unset the policy type

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+        ResourcePolicy firstResourcePolicyOfEPerson2 = ResourcePolicyBuilder.createResourcePolicy(context)
+                .withDspaceObject(collection)
+                .withAction(Constants.ADD)
+                .withPolicyType(ResourcePolicy.TYPE_CUSTOM)
```
better unset the policy type

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+      ResourcePolicy resourcePolicy = ResourcePolicyBuilder.createResourcePolicy(context)
+              .withDspaceObject(community)
+              .withAction(Constants.READ)
+              .withPolicyType(ResourcePolicy.TYPE_CUSTOM)
```
better unset the policy type

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,578 @@
+        Community community = CommunityBuilder.createCommunity(context).withName("My community").build();
+
+        ResourcePolicy resourcePolicy = ResourcePolicyBuilder.createResourcePolicy(context).withDspaceObject(community)
+                .withAction(Constants.READ).withPolicyType(ResourcePolicy.TYPE_CUSTOM).withUser(admin).build();
```
I would suggest to create a resource policy for an item related to a normal eperson other than the admin

abollini reviewed (changes requested):
Hi @Micheleboychuk 
good job! the code looks good, I have added in line some comments mainly related to javadoc and PR cleanup

Micheleboychuk committed:
added the following unit test:
	- findOneAccessGrantToAdminTest()
        - findOneAccessGrantToSomeOneUserTest()

Micheleboychuk committed:
created:
	Class:
		- DSpaceObjectRestPermissionEvaluatorPlugin
		- DSpaceObjectUtils - this class contain the method that permit to find some DSpaceObject by UUID

tdonohue commented on dspace-api/src/main/java/org/dspace/authorize/service/ResourcePolicyService.java:
```diff
@@ -66,26 +67,60 @@
+
+    public int searchCountEPerson(Context context, EPerson eperson) throws SQLException;
+
+    public List<ResourcePolicy> searchByEPersonAndResourceUuid(Context context, EPerson ePerson, UUID resourceUuid,
```
I would prefer it if we renamed these various `searchBy*` methods to be `findBy*`.  We tend to use `findBy` when searching for objects (see for example the other `findBy` methods in this same class.

tdonohue commented on dspace-api/src/main/java/org/dspace/authorize/service/ResourcePolicyService.java:
```diff
@@ -66,26 +67,60 @@
+
+    public List<ResourcePolicy> findByEPerson(Context c, EPerson ePerson, int offset, int limit) throws SQLException;
+
+    public int searchCountEPerson(Context context, EPerson eperson) throws SQLException;
```
Similar to my comment about renaming `searchBy` to `findBy`, these `searchCount` methods could probably be renamed as `count` or `countBy` methods (they don't need the word "search").  See for example, other `count` methods in other Services (e.g. `ItemService`)

tdonohue commented on dspace-api/src/main/java/org/dspace/authorize/dao/ResourcePolicyDAO.java:
```diff
@@ -79,4 +80,33 @@
+
+    public List<ResourcePolicy> findbyEPerson(Context c, EPerson ePerson, int offset, int limit) throws SQLException;
+
+    public int searchCountEPerson(Context context, EPerson eperson) throws SQLException;
```
Same here, `searchCount` could be renamed to `count` or `countBy` (throughout this class)

tdonohue commented on dspace-api/src/main/java/org/dspace/authorize/dao/ResourcePolicyDAO.java:
```diff
@@ -79,4 +80,33 @@
+
+    public int searchCountEPerson(Context context, EPerson eperson) throws SQLException;
+
+    public List<ResourcePolicy> searchByEPersonAndResourceUuid(Context context, EPerson ePerson, UUID resourceUuid,
```
And again, `searchBy` should be `findBy` to align with other DAO classes in DSpace.

tdonohue reviewed (commented):
@Micheleboychuk : Thanks for this work. Overall it's a good start! I gave this a quick review today, and I agree with @abollini 's existing notes.  I've also added a few other inline comments/suggestions below.

Micheleboychuk committed:
Implement community feedbacks

benbosman reviewed (commented):
I've verified the state of this PR, but it seems to be a bit soon to review in detail. My initial findings are:
* The URL seems to be `resourcePolicies` instead of `resourcepolicies` in the HAL browser
* The [Single Resource Policy](https://github.com/DSpace/Rest7Contract/blob/master/resourcepolicies.md#single-resource-policy) currently contains a groupUUID property instead of the expected HAL links. It only works when using uppercase P in `resourcePolicies`
* The [Search by resource](https://github.com/DSpace/Rest7Contract/blob/master/resourcepolicies.md#resource) doesn't work for me. `/server/api/authz/resourcePolicies/search/resource?uuid=2f4ec582-109e-4952-a94a-b7d7615a8c69` returns org.dspace.app.rest.exception.RepositorySearchNotFoundException and `/server/api/authz/resourcepolicies/search/resource?uuid=2f4ec582-109e-4952-a94a-b7d7615a8c69` returns org.dspace.app.rest.exception.RepositoryNotFoundException

Micheleboychuk committed:
Implemented unit test  for method createAndReturn

Micheleboychuk committed:
added static method delete by id in ResourcePolicyBuilder

Micheleboychuk committed:
added method createAndReturn

Micheleboychuk committed:
add Jackson(Access.OnlyRead) property to method getType, and fix method createAndReturn

Micheleboychuk committed:
implemented unit test deleteOne

Micheleboychuk committed:
implemented method delete()

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:
```diff
@@ -0,0 +1,577 @@
+
+
+    @Test
+    public void searchOneResoucesPolicyByEpersonUuidTest() throws Exception {
```
better to keep the test names aligned with the repository names. Replace search with find

Micheleboychuk committed:
Implement community feedbacks

Micheleboychuk committed:
Merge branch 'master' of https://github.com/DSpace/DSpace into resourcepolicies

Micheleboychuk committed:
Implemented unit tests for methods create and delete

Micheleboychuk committed:
add property @PreAuthorize for methods createAndReturn and delete

abollini committed:
Merge branch 'resourcepolicy_auth' into resourcepolicies

abollini committed:
Draft patch operations for ResourcePolicies

Micheleboychuk committed:
Merge branch 'resourcepolicies' of https://github.com/abollini/dspace into resourcepolicy_auth

merge resourcepolicy with resourcepolicy_auth

Micheleboychuk committed:
Add ITs for patch operations, fix cast errors

abollini committed:
Implement add, replace, delete operations for startDate

Micheleboychuk committed:
Added permission evaluator plagin for patch in ResourcePolicy

Micheleboychuk committed:
added methods withEndDate(), withDescriptions, withName for ResourcePolicyBuilder

Micheleboychuk committed:
added unit tests for patch

Micheleboychuk committed:
added patchOperatins for startDate and endDate

Micheleboychuk committed:
added patchOperations for name and description

abollini reviewed (approved):
Thanks @Micheleboychuk for all the hard work here. I have reviewed the code and put some comments inline about minor cleanups/improvements. I have also tested the final result and also my manual test succeed, anyway the huge amount of tests included should be a good guarantee! :)

Please take the time to merge the latest changes from master as I noted that a recent update of spring requires change in the pagination methods signature from int to long for offset, etc.

I'm going to update the description of the PR to reflect the current status.

@benbosman please note that the groupUUID and epersonUUID attributes are left in the ResourcePolicyRest class as they were used in other (unrelated part of the code base) and I suggested to Michele to deal with them in a future PR to avoid to increase the size of this PR (and also because they are likely to imply changes on the angular side)

Micheleboychuk committed:
renamed the class and add javadoc

Micheleboychuk committed:
added javadoc

Micheleboychuk committed:
Implement community feedbacks

Micheleboychuk committed:
Merge branch 'master' of https://github.com/DSpace/DSpace into resourcepolicies

Micheleboychuk committed:
added javadoc

Micheleboychuk committed:
Implemented unit test

abollini commented:
I have created https://jira.lyrasis.org/browse/DS-4418 for the eperson/group UUID attributes to remove in a future PR and flagged all my previous resolved feedback.
Thanks again for your effort on that @Micheleboychuk there are only two minors feedback residual from me. The PR is ready to be reviewed by other

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +95,211 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
+
+    @PreAuthorize("hasPermission(#resourceUuid, 'dspaceObject', 'ADMIN')")
+    @SearchRestMethod(name = "resource")
+    public Page<ResourcePolicyRest> resource(@Parameter(value = "uuid", required = true) UUID resourceUuid,
```
Could you add JavaDocs to this new method?  This also looks like it should be named something like `findByResource` as it's returning all ResourcePolicies that match a given Resource.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +95,211 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
+
+    @PreAuthorize("hasPermission(#epersonUuid, 'EPERSON', 'READ')")
+    @SearchRestMethod(name = "eperson")
+    public Page<ResourcePolicyRest> eperson(@Parameter(value = "uuid", required = true) UUID epersonUuid,
```
Again, could you add JavaDocs to this new method? This also looks like it should be named something like `findByEPerson` as it's returning all ResourcePolicies that match a given EPerson.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +95,211 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
+
+    @PreAuthorize("hasPermission(#groupUuid, 'GROUP', 'READ')")
+    @SearchRestMethod(name = "group")
+    public Page<ResourcePolicyRest> group(@Parameter(value = "uuid", required = true) UUID groupUuid,
```
Again, should have Javadocs.  Again, I'd recommend naming this `findByGroup` as it's finding ResourcePolicies that match a given Group.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/security/DSpaceObjectRestPermissionEvaluatorPlugin.java:
```diff
@@ -0,0 +1,68 @@
+@Component
+public class DSpaceObjectRestPermissionEvaluatorPlugin extends RestObjectPermissionEvaluatorPlugin {
+
+    private static final Logger log = LoggerFactory.getLogger(ResourcePolicyRestPermissionEvaluatorPlugin.class);
```
This logger is still referencing the wrong class.  As noted above, should be `DSpaceObjectRestPermissionEvaluatorPlugin.class`

tdonohue reviewed (commented):
@Micheleboychuk and @abollini : Gave this a quick code review.  Overall the code is looking good, but a few minor comments (methods that are missing JavaDocs or are misnamed) below.  I haven't had a chance to test this yet but will try to make time for it sometime this week.

Micheleboychuk committed:
added javadoc and rename some methods

Micheleboychuk committed:
Implement community feedbacks

benbosman reviewed (changes requested):
I've verified the state of this PR again, but I noticed not all my previous feedback was resolved
* The [Single Resource Policy](https://github.com/DSpace/Rest7Contract/blob/master/resourcepolicies.md#single-resource-policy) still contains a groupUUID property. The expected HAL links are present as well though
* The [Single Resource Policy](https://github.com/DSpace/Rest7Contract/blob/master/resourcepolicies.md#single-resource-policy) doesn't return the resource

I will continue my review at a later time, but these are important basic features which will make testing very hard

tdonohue reviewed (commented):
@Micheleboychuk and @abollini : Thanks for the continued work on this.  I've reviewed this again today.  The code itself looks fine, but my tests don't seem to be returning the results I'd expect.  Here's the issues that I'm seeing:

1. As @benbosman mentioned, when using `/api/authz/resourcepolicies/<:id>` for a valid policy, the embedded `resource` is always `null`.  Additionally `/api/authz/resourcepolicies/<:id>/resource` always seems to return `204 nocontent`.
2. Also, like @benbosman, for that `/api/authz/resourcepolicies/<:id>` endpoint, I'm always seeing a `groupUUID` property returned. This is really odd (as not all policies have a Group), and makes me wonder if this `groupUUID` field is actually the missing `resource`?
3. Possibly related to the first issue, if I search by resource (`/api/authz/resourcepolicies/search/resource?uuid=<:uuid>`), I seem to always get back the **same 20 resource policies** (no matter what `uuid` I pass in). For all 20 of those resource policies, the `resource` is always `null`.
4. If I customize the page size to something *less than 20*, then the number of `totalElements` changes to match the `size` (so that `totalPages=1` at all times). This seems to imply that pagination is not working in search results.  For example, `/api/authz/resourcepolicies/search/resource?uuid=<:uuid>&size=10` returns `totalPages=1` and `totalElements=10`. 
 But, `/api/authz/resourcepolicies/search/resource?uuid=<:uuid>` returns `totalPages=1` and `totalElements=20`

The bugs I'm seeing are all around the *resource*.  If I do searches based on Group or EPerson, those seem to work fine (and pagination works too).

Micheleboychuk committed:
added unit test for method findByResource

Micheleboychuk committed:
added some checks in resourcePolicyMatcher

Micheleboychuk committed:
remove references to default bitstream policies to avoid infinite loop

Micheleboychuk committed:
fix findByResourceUuid query

Micheleboychuk committed:
uniform naming and code cleanup

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +95,249 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
+     * @return It returns the list of explicit matching resource policies, no inherited or broader resource policies
+     *         will be included in the list nor policies derived by groups' membership
+     */
+    @PreAuthorize("hasPermission(#groupUuid, 'GROUP', 'READ')")
```
I think this should likely be checking for GROUP **WRITE** (or maybe even ADMIN) permissions.  

I'm noticing that while *unauthenticated*, I can view **all the resource policies** of the Anonymous group via this REST call: `http://localhost:8080/server/api/authz/resourcepolicies/search/group?uuid=[uuid-of-Anonymous]` 

This seems like a security issue to me.

abollini replied:
We have had a similar discussion during the definition of the REST contract see https://github.com/DSpace/Rest7Contract/pull/87#discussion_r343754096
essentially we allow people to know what they are allowed to do.

The current contract says that the search method is available to member of the group and this actually mean READ permission. Maybe in future we could find an use case to give READ permission over a group without be a member of such group but this is not a feature that exists right now.
If you prefer to deal with this difference we should hardcode this check in the repository search method as it is not possible to express the "isMemberOf" concept as a DSpaceRestPermission (READ, WRITE, etc)

abollini replied:
this is the related test https://github.com/DSpace/DSpace/pull/2604/files#diff-3a5daad2fc92c374f068f92a31c36ec7R857

tdonohue replied:
@abollini : I understand your point.  However, I'm still a little concerned about making *every policy* related to the `Anonymous` group **publicly readable.**  I wonder if we should leave this `@PreAuthorize` as-is, but implement a special (hardcoded) check for the Anonymous group (to ensure it is only visible to authenticated users, perhaps)?  I don't like the idea of having any resource policies being publicly readable....at the minimum you need to have an authenticated account in the system.

abollini replied:
ok, implemented. A test has been added for this use case
https://github.com/DSpace/DSpace/pull/2604/commits/434774f1f650683f87366c26b1a373aa18a9c64c

it was failing with the original code https://travis-ci.org/DSpace/DSpace/builds/646783645?utm_source=github_status&utm_medium=notification
and fixed now  https://travis-ci.org/DSpace/DSpace/builds/646784796?utm_source=github_status&utm_medium=notification

Here the PR to sync the contract too https://github.com/DSpace/Rest7Contract/pull/101

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ResourcePolicyRestRepository.java:
```diff
@@ -52,9 +95,249 @@ public ResourcePolicyRest findOne(Context context, Integer id) {
+     * @return It returns the list of explicit matching resource policies, no inherited or broader resource policies
+     *         will be included in the list nor policies derived by groups' membership
+     */
+    @PreAuthorize("hasPermission(#epersonUuid, 'EPERSON', 'READ')")
```
Same here, I think we should require WRITE or ADMIN permissions here.  READ may not be tight enough.

abollini replied:
as above, here the test to verify that user can only access policies that belong to themself
https://github.com/DSpace/DSpace/pull/2604/files#diff-3a5daad2fc92c374f068f92a31c36ec7R393

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/security/DSpaceObjectRestPermissionEvaluatorPlugin.java:
```diff
@@ -0,0 +1,72 @@
+ * @author Mykhaylo Boychuk (4science.it)
+ */
+@Component
+public class DSpaceObjectRestPermissionEvaluatorPlugin extends RestObjectPermissionEvaluatorPlugin {
```
This PermissionEvaluatorPlugin seems oddly named to me. It looks like it's evaluating for *Admin* permissions on a DSpaceObject. Shouldn't it be named `DSpaceObjectAdminPermissionEvaluatorPlugin` (or something similar, just like the `ResourcePolicyAdminPermissionEvaluatorPlugin`)?

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/DSpaceObjectUtils.java:
```diff
@@ -0,0 +1,43 @@
+    @Autowired
+    private ContentServiceFactory contentServiceFactory;
+
+    public DSpaceObject findDSpaceObject(Context context, UUID uuid) throws SQLException {
```
This method should have Javadocs, please

abollini replied:
done, I don't know why github doesn't note it https://github.com/DSpace/DSpace/pull/2604/commits/03238bcbb5fd5303dc7b7a7a1d696096ad935391#diff-26c81a6c14a0b2489296ceec4d1dc488R33

tdonohue reviewed (commented):
@Micheleboychuk and @abollini : Thanks for the updates.  Overall, this is looking very good now!

I've retested this today, and *all of the previously reported bugs* look to be fixed. I no longer can reproduce them.

I've also done an updated code review and found a few, minor issues to resolve (see below inline comments).  After that, I think this is ready to merge.

**UPDATE**: I did just notice a (very minor) oddity in `/api/authz/resourcepolicies/<:id>`. ResourcePolicies seem to return either a `epersonUUID` or `groupUUID` based on whether they are an EPerson based or Group based policy.  This was unexpected, as the EPerson or Group is already referenced in `_links` and `_embedded`.  You can see this most easily by using the search by Eperson or Group.

abollini commented:
@tdonohue please see my comments about the preAuthorize check and let us know if it is fine for now to stay with the READ check or if we should implement a real isMemberOf check as stated in the contract.
About

> UPDATE: I did just notice a (very minor) oddity in /api/authz/resourcepolicies/<:id>. ResourcePolicies seem to return either a epersonUUID or groupUUID based on whether they are an EPerson based or Group based policy. This was unexpected, as the EPerson or Group is already referenced in _links and _embedded. You can see this most easily by using the search by Eperson or Group.

Please note that this is tracked in https://jira.lyrasis.org/browse/DS-4418 and, as it is a larger unrelated issue than the original goal of this PR (the attributes were already here) and it can have an impact on the angular side I'm proposing to deal with it in a separate PR

abollini committed:
Community feedback: improved javadoc and naming

Micheleboychuk committed:
Add test to verify that anonymous users cannot search by group

Micheleboychuk committed:
Forbid anonymous users to search by gorup

benbosman reviewed (commented):
I've tested the read-only functionality again, and noticed one last problem

When requesting a resource policy attached to the Anonymous group without using authentication, I receive 401 Unauthorized which doesn't correspond to https://github.com/DSpace/Rest7Contract/blob/master/resourcepolicies.md#single-resource-policy

I will also test with a user with limited permissions and to create, update, delete resource policies

abollini commented:
Hi @benbosman 
thanks for your review. You are right and we are including a test to reproduce what you have reported.
Personally I would stick to the current contract but I want just to double check if @tdonohue instead doesn't prefer to restrict also this endpoint to loggedin user as requested for the search byGroup method. In such case we will fix the "new" coming test to reflect this decision as well

Micheleboychuk committed:
added unit test to retrieve a resource policy attached to the Anonymous group with/without using authentication

tdonohue commented:
@benbosman and @abollini : I'd prefer that all resource policy endpoints *require* authentication (at a minimum).  Resource Policies obviously describe the permissions on an object in DSpace.  It seems like a security issue to allow anonymous users to easily see what permissions exist on specific  objects, as it potentially opens some objects up to hacking if the wrong permissions are accidentally attached.   

For example, imagine if an Item accidentally had a policy which allowed for Anonymous EDIT permissions....and that was easily discoverable to any anonymous user via the REST API. Anyone could then easily post a request via the REST API to modify that Item.

Micheleboychuk committed:
update test to access single resource policy owned by Anonymous group

resource policy endpoints are now restricted to authenticated people according to the reviewed contract

abollini commented:
@tdonohue @benbosman it should be in the final stage now. We have updated the PR to reflect the agreed decision to limit the endpoints to loggedin users. The contract PR is ready as well https://github.com/DSpace/Rest7Contract/pull/101

tdonohue reviewed (approved):
👍 Retested today, and none of the endpoints are available anonymously anymore.  Recent code changes also look good (and thanks for the additional tests to prove the changes work).  Thanks for your hard work on this @Micheleboychuk !

benbosman reviewed (commented):
While testing the functionality, I noticed.
* Resources are made available via /server/api/authz/resourcepolicies/<:id>/resource while they're not available via the self link of the resources (no READ permissions are verified, if any permission is present, it' accessible)

I didn't get to test creating and updating resource policies yet, but I assume this was already reviewed by Tim and Andrea in detail, and I noticed there are ITs which verify this process
I briefly read through the code, and didn't notice any immediate big problems.
I assume this was already reviewed by Tim and Andrea in detail

benbosman reviewed (approved):
Since @abollini verified the bug from https://github.com/DSpace/DSpace/pull/2604#pullrequestreview-356850703 is not specific to this PR, I'm approving the PR assuming a JIRA ticket is created

abollini commented:
yes, as anticipated via slack it is a more general issue now tracked in https://jira.lyrasis.org/browse/DS-4432 (I have also written tests to show this issue on abother endpoint see comment in the JIRA ticket)

tdonohue commented:
Merging as this is at +3 & the outstanding issue is tracked in DS-4432.
